### PR TITLE
updated training for LFC and CNV outperforming old one without any scaling

### PR DESCRIPTION
**Pull request by ussamazahid96**

### CODE REVIEW

**Comment by maltanar** [comment]
@ussamazahid96 can you provide a short bullet list summary of what change(s) are made to the networks and the training method?

**Comment by ussamazahid96** [comment]
@maltanar the following changes were applied as compared to your master branch:
**Changes in the training settings:**
1. Squared hinge loss instead of cross-entropy loss
2. Adam optimizer instead of SGD (initial LR set to 0.02, batch-size to 100, weight decay to 0)
3. Fixed LR schedule (halved after every 40 epochs)
4. Weights clipping on the underlying FP weights
5. Accuracy calculation without the topk function (which has a bug that the top1 accuracy also depends on the other topk accuracy you are evaluating (e.g. top5), so it doesn't matches with the hardware accuracy)
**Changes applied to the CNV model:**
6. Input quantization to Q1.7 format using the QuantHardTanh function before the first conv layer, as the hardware uses.
7. Max-pool placed after the activation layer rather than conv (it is strange but PyTorch gives many negative values in Batchnorm2d which on hardware reflects as a min pool operation when you change the order of Max-pool operation in hardware, whereas Theano never gives negative values in Batchnorm2d so it is always a max pool even if you change the order on hardware)
8. Last BatchNorm at the end, which uses 1 value for mean, invstd, beta and gamma, instead of 10.
9. Weights init from uniform
10. All scaling disabled (scale const set to 1) and narrow range enabled for all layers

**Comment by maltanar** [comment]
Thanks! I've been working on switching to your versions of these networks for FINN, although I had to make further changes -- you can see those on  the feature/rework_scaling_clipping branch. Your PR will be merged in the next FINN release.
